### PR TITLE
chore: rework base starter project to work with latest OSGi support in Flow

### PR DESCRIPTION
**Pull request by denis-anisimov**
base starter project targets now Flow master `6.0` with `19.0` Platform

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/base-starter-flow-osgi/98)

### CODE REVIEW

File: starter/pom.xml [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>com.vaadin</groupId>
+                <artifactId>vaadin-bom</artifactId>
+                <type>pom</type>
+                <scope>import</scope>
+                <version>${vaadin.version}</version>
+            </dependency>
```
Comment (pleku) [inline starter/pom.xml:27]: This is already in the parent project, so do we need it here too ?

**Review by pleku — CHANGES_REQUESTED** [review]

File: app/pom.xml [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        <dependency>
+            <groupId>com.example</groupId>
+            <artifactId>project-base-osgi</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        
+        <dependency>
+            <groupId>com.vaadin</groupId>
+            <artifactId>vaadin-button-flow</artifactId>
+        </dependency>
```
Comment (pleku) [inline app/pom.xml:31]: A dumb question - why is this here ?

**Comment by pleku** [comment]
When I run the project with the readme recommended (AdoptOpenJDK Java 11)
```
mvn clean install
java -jar app/target/app.jar
```
I get 500 error
```
[pool-2-thread-1] INFO com.vaadin.flow.server.startup.VaadinAppShellInitializer - Using com.example.starter.base.AppShell class for configuring `index.html` response
[qtp446445803-45] ERROR com.vaadin.flow.server.DefaultErrorHandler -
java.lang.NullPointerException
        at com.vaadin.flow.server.frontend.FrontendUtils.getFileFromWebpack(FrontendUtils.java:631)
        at com.vaadin.flow.server.frontend.FrontendUtils.getFileContent(FrontendUtils.java:477)
        at com.vaadin.flow.server.frontend.FrontendUtils.getIndexHtmlContent(FrontendUtils.java:466)
        at com.vaadin.flow.server.communication.IndexHtmlRequestHandler.getIndexHtmlDocument(IndexHtmlRequestHandler.java:225)
        at com.vaadin.flow.server.communication.IndexHtmlRequestHandler.synchronizedHandleRequest(IndexHtmlRequestHandler.java:70)
        at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40)
        at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1567)
        at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:278)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
        at org.apache.felix.http.base.internal.handler.ServletHandler.handle(ServletHandler.java:123)
        at org.apache.felix.http.base.internal.dispatch.InvocationChain.doFilter(InvocationChain.java:86)
        at org.apache.felix.http.base.internal.dispatch.Dispatcher$1.doFilter(Dispatcher.java:146)
        at org.apache.felix.http.base.internal.whiteboard.WhiteboardManager.invokePreprocessors(WhiteboardManager.java:988)
        at org.apache.felix.http.base.internal.dispatch.Dispatcher.dispatch(Dispatcher.java:91)
        at org.apache.felix.http.base.internal.dispatch.DispatcherServlet.service(DispatcherServlet.java:49)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:569)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1610)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1377)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:507)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1580)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1292)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:501)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
        at java.base/java.lang.Thread.run(Thread.java:834)
[ERROR] Exception while processing request to /
javax.servlet.ServletException: com.vaadin.flow.server.ServiceException: java.lang.NullPointerException
        at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:280)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
        at org.apache.felix.http.base.internal.handler.ServletHandler.handle(ServletHandler.java:123)
        at org.apache.felix.http.base.internal.dispatch.InvocationChain.doFilter(InvocationChain.java:86)
        at org.apache.felix.http.base.internal.dispatch.Dispatcher$1.doFilter(Dispatcher.java:146)
        at org.apache.felix.http.base.internal.whiteboard.WhiteboardManager.invokePreprocessors(WhiteboardManager.java:988)
        at org.apache.felix.http.base.internal.dispatch.Dispatcher.dispatch(Dispatcher.java:91)
        at org.apache.felix.http.base.internal.dispatch.DispatcherServlet.service(DispatcherServlet.java:49)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763)
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:569)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1610)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1377)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:507)
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1580)
        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1292)
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
        at org.eclipse.jetty.server.Server.handle(Server.java:501)
        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.vaadin.flow.server.ServiceException: java.lang.NullPointerException
        at com.vaadin.flow.server.VaadinService.handleExceptionDuringRequest(VaadinService.java:1616)
        at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1579)
        at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:278)
        ... 33 more
Caused by: java.lang.NullPointerException
        at com.vaadin.flow.server.frontend.FrontendUtils.getFileFromWebpack(FrontendUtils.java:631)
        at com.vaadin.flow.server.frontend.FrontendUtils.getFileContent(FrontendUtils.java:477)
        at com.vaadin.flow.server.frontend.FrontendUtils.getIndexHtmlContent(FrontendUtils.java:466)
        at com.vaadin.flow.server.communication.IndexHtmlRequestHandler.getIndexHtmlDocument(IndexHtmlRequestHandler.java:225)
        at com.vaadin.flow.server.communication.IndexHtmlRequestHandler.synchronizedHandleRequest(IndexHtmlRequestHandler.java:70)
        at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40)
        at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1567)
        ... 34 more
```

Same happens for the individual steps
```
mvn -pl starter install
mvn -pl app bnd-resolver:resolve
mvn -pl app bnd-run:run
```

**Comment by pleku** [comment]
But if I run 
```
mvn install -Pproduction
java -jar app/target/app.jar
```
then it works. 

@denis-anisimov shouldn't we make it so that `build-frontend` is always executed for the `starter` project and then the production profile only would only add the production mode artifact (no debug information) ?

**Comment by denis-anisimov** [comment]
> ```
> 
> then it works.
> 
> @denis-anisimov shouldn't we make it so that `build-frontend` is always executed for the `starter` project and then the production profile only would only add the production mode artifact (no debug information) ?

`build-frontend` should be always executed for the `starter` , yes.
I thought I've done already it in this way.
I will check.

File: app/pom.xml (hunk 2 above)
Comment (denis-anisimov) [reply app/pom.xml:31]: The question is good.
It's a leftover from the first attempt to make the starter work without 19 Platform.
Now it's not needed anymore.

File: starter/pom.xml (hunk 1 above)
Comment (denis-anisimov) [reply starter/pom.xml:27]: Not needed, indeed.
Removed.

File: pom.xml [hunk 3]
```diff
@@ -59,446 +54,43 @@
...
-            <groupId>com.vaadin</groupId>
-            <!-- Replace artifactId with vaadin-core to use only free components -->
-            <artifactId>vaadin</artifactId>
-        </dependency>
-
-   
         <dependency>
             <groupId>com.vaadin</groupId>
             <artifactId>flow-osgi</artifactId>
+            <version>6.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (denis-anisimov) [reply pom.xml:62]: Leftovers, right.

**Comment by denis-anisimov** [comment]
> > ```
> > 
> > then it works.
> > 
> > @denis-anisimov shouldn't we make it so that `build-frontend` is always executed for the `starter` project and then the production profile only would only add the production mode artifact (no debug information) ?
> > ```
> 
> `build-frontend` should be always executed for the `starter` , yes.
> I thought I've done already it in this way.
> I will check.

I cannot reproduce this.
And this is what I expect.
`production` profile is set active by default:
```
<activation>
                <activeByDefault>true</activeByDefault>
            </activation>
```

I think I've seen also some weird behavior when token file has not been added into the starter artifact.
But didn't realize why it happened.
Now if I remove `starter/target` and then run `mvn clean install` the token file is generated :`starter/target/classes/META-INF/VAADIN/config/flow-build-info.json`.
And I see it's added into the resulting jar file.

Do you have for some unknown reason the `production`  profile inactive by default ?
So is your `pom.xml` in the starter project different than it's in this PR?

**Comment by pleku** [comment]
>Do you have for some unknown reason the production profile inactive by default ?
So is your pom.xml in the starter project different than it's in this PR?

My guess is that it is because I'm running with Java 11 - there is another profile for that with `<activation><jdk>11</jdk></activation>`, and apparently when that activates then the `<activeByDefault>true</activeByDefault>` profile is not run.

I think I would just move the `<goal>build-frontend</goal>` to the default vaadin-maven-plugin configuration and leave the `production` profile to not be active by default, but it would only have the `flow-server-production-mode` artifact in it.

**Comment by pleku** [comment]
Yes, this behavior is an Maven feature:
```All profiles that are active by default are automatically deactivated when a profile in the POM is activated on the command line or through its activation config.```

**Comment by denis-anisimov** [comment]
> Yes, this behavior is a Maven feature:
> `All profiles that are active by default are automatically deactivated when a profile in the POM is activated on the command line or through its activation config.`

OK, quite unexpected for me.
Moved goal into the main config section.

File: starter/pom.xml [hunk 4]
```diff
@@ -166,41 +155,6 @@
     </build>
 
     <profiles>
```
Comment (pleku) [inline starter/pom.xml:157]:
```suggestion
    <profiles>
        <profile>	
            <!-- Vaadin Production mode is activated using -Pproduction -->	
            <id>production</id>	
            <activation>	
                <activeByDefault>true</activeByDefault>	
            </activation>	
            <properties>	
                <vaadin.productionMode>true</vaadin.productionMode>	
            </properties>	
             <dependencies>	
                <dependency>	
                    <groupId>com.vaadin</groupId>	
                    <artifactId>flow-server-production-mode</artifactId>	
                </dependency>	
            </dependencies>
        </profile>
```

**Review by pleku — CHANGES_REQUESTED** [review]
I would still add keep the production profile for enabling the production mode for flow-server

File: starter/pom.xml (hunk 4 above)
Comment (denis-anisimov) [reply starter/pom.xml:157]: I'm pretty sure it doesn't work on OSGi.
`flow-server-production-mode` contains only one resource `web-fragment.xml`  and I don't see any way how it may work in OSGi.

In WAR: the fragment is a part of WAR file , so it's clear which application to apply the fragment to.

In OSGi this jar "will be loaded" as  a standalone bundle and there is no any way to associate the web app bundle with the fragment.

In fact "will be loaded" is false. `flow-server-production-mode` has no OSGi compatible manifest and it won't be deployed at all.
So this config has no any effect in OSGi for many various reasons.
production mode can be enabled via `VaadinMode` annotation.
I will add the parameter to it.

Comment (pleku) [reply starter/pom.xml:157]: Right ... I think it will be fine then

**Review by pleku — APPROVED** [review]

File: starter/pom.xml (hunk 4 above)
Comment (denis-anisimov) [reply starter/pom.xml:157]: Added `productionMode` parameter to the `VaadinMode` annotation.

**Review by pleku — APPROVED** [review]

**Review by pleku — APPROVED** [review]
